### ISSUE TITLE
99% cpu and never ends on 1st bytes of UTF data as pattern

### ISSUE THREAD

**Issue by korziner** [Genivia/ugrep#224]
cut -b1 *0.00936853889674682*sql.fingerprint3alpha|paste -s|tr -d \\t
s�p�p�p�p�p�p�p�v�p�lfdw�
cut -b1-4 *0.00936853889674682*sql.fingerprint3alpha
sele
ти
post
ти
post
ти
post
ти
post
ти
pmUs
ти
post
ти
pmUs
ти
vers
ти
post
ти
...
```
cut -b1 *0.00936853889674682*sql.fingerprint3alpha|paste -s|tr -d \\t|sudo operf `which ugrep` -Z1 $(cut -b1 *0.00936853889674682*sql.fingerprint3alpha|paste -s|tr -d \\t)
operf: Profiler started
^C* * * * WARNING: Profiling rate was throttled back by the kernel * * * *
The number of samples actually recorded is less than expected, but is
probably still statistically valid.  Decreasing the sampling rate is the
best option if you want to avoid throttling.

Profiling done.
opreport -cl --demangle=smart `which ugrep`
Using ../PGmd5uniq/oprofile_data/samples/ for samples directory.

WARNING! Some of the events were throttled. Throttling occurs when
the initial sample rate is too high, causing an excessive number of
interrupts.  Decrease the sampling frequency. Check the directory
../PGmd5uniq/oprofile_data/samples/current/stats/throttled
for the throttled event names.

warning: /ath10k_core could not be found.
warning: /ip_tables could not be found.
warning: /mac80211 could not be found.
CPU: Intel Skylake microarchitecture, speed 2300 MHz (estimated)
Counted cpu_clk_unhalted events () with a unit mask of 0x00 (Core cycles when at least one thread on the physical core is not in halt state) count 100000
samples  %        image name               symbol name
-------------------------------------------------------------------------------
31499    41.2577  libc-2.27.so             __strchr_sse2
  31499    100.000  libc-2.27.so             __strchr_sse2 [self]
-------------------------------------------------------------------------------
18205    23.8451  ugrep                    reflex::convert(char const*, char const*, int, map<__cxx11::string, __cxx11::string> const*)
  18205    100.000  ugrep                    reflex::convert(char const*, char const*, int, map<__cxx11::string, __cxx11::string> const*) [self]
-------------------------------------------------------------------------------
10286    13.4727  libstdc++.so.6.0.29      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.29
  10286    100.000  libstdc++.so.6.0.29      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.29 [self]
-------------------------------------------------------------------------------
6626      8.6788  ugrep                    reflex::is_modified(map<unsigned long, __cxx11::string, less<unsigned long>, allocator<pair<unsigned long const, __cxx11::string>>> const&, int)
  6626     100.000  ugrep                    reflex::is_modified(map<unsigned long, __cxx11::string, less<unsigned long>, allocator<pair<unsigned long const, __cxx11::string>>> const&, int) [self]
-------------------------------------------------------------------------------
6512      8.5295  ugrep                    reflex::supports_escape(char const*, int)
  6512     100.000  ugrep                    reflex::supports_escape(char const*, int) [self]
-------------------------------------------------------------------------------
3198      4.1888  libc-2.27.so             isalpha
  3198     100.000  libc-2.27.so             isalpha [self]
-------------------------------------------------------------------------------
6         0.0079  ath10k_core              /ath10k_core
  6        100.000  ath10k_core              /ath10k_core [self]
-------------------------------------------------------------------------------
6         0.0079  mac80211                 /mac80211
  6        100.000  mac80211                 /mac80211 [self]
-------------------------------------------------------------------------------
5         0.0065  ld-2.27.so               do_lookup_x
  5        100.000  ld-2.27.so               do_lookup_x [self]
-------------------------------------------------------------------------------
2         0.0026  ld-2.27.so               _dl_relocate_object
  2        100.000  ld-2.27.so               _dl_relocate_object [self]
-------------------------------------------------------------------------------
1         0.0013  ip_tables                /ip_tables
  1        100.000  ip_tables                /ip_tables [self]
-------------------------------------------------------------------------------
1         0.0013  ld-2.27.so               _dl_lookup_symbol_x
  1        100.000  ld-2.27.so               _dl_lookup_symbol_x [self]
-------------------------------------------------------------------------------
``` 
ugrep -V
ugrep 3.9.1 x86_64-pc-linux-gnu +sse2 +pcre2jit +zlib +bzip2 +lzma +lz4 +zstd

**Comment by genivia-inc** [comment]
Looks like a potential problem with the regex pattern passed to ugrep, not with the search itself. The operf output is not helpful to see what it could be. If it keeps running in `reflex::convert` then it has something to do with the pattern that `convert` analyzes. I suppose this pattern is quite large? I doubt this has anything to do with `-Z1`, but I could be wrong.

Try the same run without `-Z1` then try with `-F` to match strings.

**Comment by korziner** [comment]
Thanks for early reply. It is not quite large 
```
cut -b1-2 UTF.txt|paste -s|tr -d \\t|~/ugrep -F $(cut -b1-2 UTF.txt|paste -s|tr -d \\t) 
seтpoтpoтpoтpoтpmтpoтpmтveтpoтlpfrdbwhт
cut -b1 UTF.txt|paste -s|tr -d \\t|~/ugrep -F $(cut -b1 UTF.txt|paste -s|tr -d \\t) 
Binary file (standard input) matches
cut -b1 UTF.txt|paste -s|tr -d \\t|~/ugrep -Z1 $(cut -b1 UTF.txt|paste -s|tr -d \\t) 
^C
cut -b1 UTF.txt|paste -s|tr -d \\t|~/ugrep -Fa $(cut -b1 UTF.txt|paste -s|tr -d \\t) 
s�p�p�p�p�p�p�p�v�p�lfdw�
```

**Comment by genivia-inc** [comment]
Thanks, this helps. What happens without `-Z1` and `-F`? Or try with `-U` to disable Unicode, so this may give some clue as to what could be a possible cause? Also, what is this UTF.txt file?

This should be easy to fix in the pattern analyzer, if I can replicate the problem with a copy of UTF.txt. Something goes wrong in the pattern analysis phase in `reflex::convert` where it appears to get stuck is my best guess.

**Comment by genivia-inc** [comment]
I don't think I'm able to reproduce this without the UTF.txt file you're using. Please attach the UTF.txt file.

**Comment by korziner** [comment]
save this as UTF.txt ```
sele
ти
post
ти
post
ти
post
ти
post
ти
pmUs
ти
post
ти
pmUs
ти
vers
ти
post
ти
```

**Comment by genivia-inc** [comment]
Found the problem and fixed it in the reflex::convert regex analyzer. This happened when Unicode regex matching is enabled and the regex pattern includes an erroneous UTF-8 character sequence and some other conditions are met.

**Comment by genivia-inc** [comment]
Fixed in v3.9.2.

**Comment by genivia-inc** [comment]
Thanks for reporting this issue. It is always nice when people take the time to explain if something is not working. It does not happen often. Reporting problems is important.
